### PR TITLE
New rule S6588: Use `DateTime(Offset).UnixEpoch` instead of pointing to that date using the constructor

### LINKED ISSUE #7080: New rule S6588: Use `DateTime(Offset).UnixEpoch` instead of pointing to that date using the constructor
Implement new rule [S6588: Use DateTime(Offset).UnixEpoch instead of pointing to that date using the constructor](https://github.com/SonarSource/rspec/pull/1747).

This rule should raise when a new `DateTime`, `DateTimeOffset` or `DateOnly` object is created with Date 1-1-1970. 

- [ ] Implement Rule S6588 for C#
- [ ] Implement Rule S6588 for VB.NET

### PR DESCRIPTION
**Pull request by cristian-ambrosini-sonarsource**
Fixes #7080

### CODE REVIEW

File: analyzers/its/sources/ManuallyAddedNoncompliantIssues.VB/IntentionalFindings/S6588.vb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿Imports System
+
+Friend Class S6588
+    Private dateTime As Date = New DateTime(1970, 1, 1) ' Noncompliant (S6588)
+    Private dateTimeOffset As DateTimeOffset = New DateTimeOffset(1970, 1, 1, 1, 1, 1, TimeSpan.Zero) ' Noncompliant
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/its/sources/ManuallyAddedNoncompliantIssues.VB/IntentionalFindings/S6588.vb:5]: The Unix Epoch is `00:00:00.0000000 UTC, January 1, 1970`, so this should be:
```vb
New DateTimeOffset(1970, 1, 1, 0, 0, 0, TimeSpan.Zero) ' Noncompliant
```

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        || (expressions[0] is TMemberAccessExpression memberAccess && IsDateTimeKindUtc(memberAccess));
+
+    private bool IsCalendarNonExistingOrGregorian(IMethodParameterLookup lookup) =>
+        !lookup.TryGetSyntax("calendar", out var expressions) || IsGregorianCalendar(expressions[0]);
+
+    private bool IsOffsetNonExistingOrZero(IMethodParameterLookup lookup) =>
+        !lookup.TryGetSyntax("offset", out var expressions) || IsZeroTimeOffset(expressions[0]);
+
+    // DateTime.UnixEpoch introduced at .NET Core 2.1/.NET Standard 2.1
+    private static bool CompilationTargetsValidNetVersion(Compilation compilation) =>
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:114]: How about `IsUnixEpochProprtySupported`?

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    var argumentCount = Language.Syntax.ArgumentExpressions(c.Node).Count();
+                    if (argumentCount < 3)
+                    {
+                        return;
+                    }
+
+                    var type = c.SemanticModel.GetTypeInfo(c.Node).Type;
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:56]: This will check every constructor call with 3 or more arguments with the semantic model. To improve the performance you can first check the name of the constructor call using only the syntax tree. If it's DateTime or DateTimeOffset then proceed with the semantic check. This will have some FNs:
- when `DateTime` or `DateTimeOffset` are aliased with the using directive
- when someone introduces a new custom type called `DateTime` or `DateTimeOffset` (you already covered it in the UTs)
Both of these scenarios are very rare, and the chance of any of them occurring when the developer also calls these constructors with these specific parameters is astronomical.
It's better to document these as FNs and make the analyzer more performant.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                },
+                Language.SyntaxKind.ObjectCreationExpressions);
+        });
+
+    protected static bool IsValueEqualsTo(TLiteralExpression literal, int value) =>
+        literal.ChildTokens().First().ValueText == value.ToString();
+
+    private bool IsEpochCtor(SyntaxNode node, SemanticModel model)
+    {
+        var methodSymbol = (IMethodSymbol)model.GetSymbolInfo(node).Symbol;
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:74]: Do we need a symbolic lookup here in every case? This can be done completely with the syntax tree (even though it will have a more complex code). Or you can do a quick syntax check: does the method have at least one argument with the value of `1970` and at least two with the value of `1`? Only check the semantic mode if the answer to that is yes.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/UseUnixEpochTest.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/UseUnixEpochTest.cs:1]: Add a test case for target-typed object creation (available from C# 9.0):
`DateTime d = new(1970, 1, 1); // Noncompliant`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.Framework.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+using System.Globalization;
+
+public class Program
+{
+    private readonly DateTime Epoch = new DateTime(1970, 1, 1); // Compliant
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.Framework.cs:6]: `// Compliant - the UnixEpoch property is only available in .NET Core 2.1/.NET Standard 2.1 and newer, the .NET Framework does not support it`

**Review by zsolt-kolbay-sonarsource — CHANGES_REQUESTED** [review]
Review round 1

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs (hunk 2 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:114]: I think I prefer `CompilationTargetsValidNetVersion`. It's also consistent with a few other examples in the repo

Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:114]: Furthermore, should this be cached?

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs (hunk 4 above)
Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:74]: If you want to do that, you should add a test case with `new DateTime(1970, 01, 01)`, because otherwise you will have some false negatives.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+                    else if (type.DerivesFrom(KnownType.System_DateTimeOffset) && IsEpochCtor(c.Node, c.SemanticModel))
+                    {
+                        c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation(), "DateTimeOffset"));
+                    }
+                },
+                Language.SyntaxKind.ObjectCreationExpressions);
+        });
+
+    protected static bool IsValueEqualsTo(TLiteralExpression literal, int value) =>
+        literal.ChildTokens().First().ValueText == value.ToString();
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:76]: Based on this comment: instead of checking for string equality, use `int.TryParse()` and compare the actual integer value. Then add test cases for 01 and 0x1.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs (hunk 4 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:80]: The rule is raising correctly for `new DateTime(1970, 01, 01)`.
The syntax check on the 1/1/1970 arguments will already filter out most of the constructors, so I would keep it as is for the rest to be 100% sure the type is indeed _System.DateTime_. The overall perf cost will be very small and we will raise with more confidence.

Comment (Corniel) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:80]: @cristian-ambrosini-sonarsource I don't see a test case for it though?

Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:80]: You can find it in the `EdgeCases` method.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs (hunk 2 above)
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:120]: IMO this way of checking the framework will be replaced in the near future with a formalized way of doing it (and consistent with all the rules). Having the same temporary, yet unprecise name might help to spot the affected rules and transition to the new system. Having said that, I can update the name as you suggest since mine is not a strong opinion👍

**Comment by Corniel** [comment]
Extra remark: writing a code fix for this one should be simple, so please do.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+public class Program
+{
+    private readonly DateTime Epoch = new DateTime(1970, 1, 1); // Noncompliant {{Use "DateTime.UnixEpoch" instead of creating DateTime instances that point to the unix epoch time}}
+    //                                ^^^^^^^^^^^^^^^^^^^^^^^^
+
+    private readonly DateTimeOffset EpochOff = new DateTimeOffset(1970, 1, 1, 0, 0, 0, 0, TimeSpan.Zero); // Noncompliant {{Use "DateTimeOffset.UnixEpoch" instead of creating DateTimeOffset instances that point to the unix epoch time}}
+    //                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    void MyMethod(DateTime dateTime)
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs:12]: `Basics` or `BasicCases`?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs:1]: Add a test case for the fully qualified type name: `System.DateTime`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb:1]: Add a VB.NET test case for the different casing of the type name: datetime or DATETIME

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+    void EdgeCases()
+    {
+        var dateTimeOffset = new DateTimeOffset(new DateTime(1970, 1, 1), new TimeSpan(0, 0, 0)); // Noncompliant
+        var dateTime = new DateTime(true ? 1970 : 1971, 1, 1); // FN
+        dateTime = new DateTime(1970, 01, 01); // Noncompliant
+        dateTime = new DateTime(1970, 0x1, 0x1); // Noncompliant
+    }
+
+    void DateTimeConstructors(int ticks, int year, int month, int day, int hour, int minute, int second, int millisecond, Calendar calendar, DateTimeKind kind)
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs:39]: That's a really nice level of detail for the tests! 👍

**Review by zsolt-kolbay-sonarsource — CHANGES_REQUESTED** [review]
The implementation looks good. Now there's only some polishing is left.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb [hunk 9]
```diff
@@ -0,0 +1,194 @@
+﻿Imports System
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb:1]: `Imports System` is not needed.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+    private static bool IsLiteralAndEqualTo(SyntaxNode node, int value) =>
+        node is TLiteralExpression literal && IsValueEqualsTo(literal, value);
+
+    private bool IsDateTimeKindNonExistingOrUtc(IMethodParameterLookup lookup) =>
+        !lookup.TryGetSyntax("kind", out var expressions)
+        || (expressions[0] is TMemberAccessExpression memberAccess && IsDateTimeKindUtc(memberAccess));
+
+    private bool IsCalendarNonExistingOrGregorian(IMethodParameterLookup lookup) =>
+        !lookup.TryGetSyntax("calendar", out var expressions) || IsGregorianCalendar(expressions[0]);
```
[the remark this answers is not in the document]
Comment (cristian-ambrosini-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:114]: There is no `TryGetSymbol` overload that accepts the parameter name. I think this is enough and we can allow the FNs you mentioned above

[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseUnixEpochBase.cs:106]: I see. Those are edge cases, but you can still cover them if you introduce a helper method:
```cs
private static bool TryGetParameterSymbol(IMethodParameterLookup lookup, string parameterName, out IParameterSymbol symbol)
{
    symbol = null;
    return lookup.TryGetSyntax(parameterName, out var expressions) && lookup.TryGetSymbol(expressions[0], out symbol);
}
```
Then use it instead of `TryGetSyntax`.

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM
[There's one comment left](https://github.com/SonarSource/sonar-dotnet/pull/7509#discussion_r1246175533), but it's an edge case, so check it if you want to address it or not.

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/UseUnixEpochTest.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+public class UseUnixEpochTest
+{
+    private readonly VerifierBuilder builderCS = new VerifierBuilder<CS.UseUnixEpoch>();
+    private readonly VerifierBuilder builderVB = new VerifierBuilder<VB.UseUnixEpoch>();
+
+#if NETFRAMEWORK
+
+    [TestMethod]
+    public void UseUnixEpoch_Framework_CS() =>
+        builderCS.AddPaths("UseUnixEpoch.Framework.cs").Verify();
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/UseUnixEpochTest.cs:36]:
```suggestion
        builderCS.AddPaths("UseUnixEpoch.Framework.cs").VerifyNoIssueReported();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/UseUnixEpochTest.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+    [TestMethod]
+    public void UseUnixEpoch_Framework_VB() =>
+        builderVB.AddPaths("UseUnixEpoch.Framework.vb").Verify();
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/UseUnixEpochTest.cs:40]:
```suggestion
        builderVB.AddPaths("UseUnixEpoch.Framework.vb").VerifyNoIssueReported();
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs [hunk 13]
```diff
@@ -0,0 +1,211 @@
+﻿using System;
+using System.Globalization;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.cs:2]: please add a test case with aliasing `DateTime`

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb [hunk 14]
```diff
@@ -0,0 +1,194 @@
+﻿Imports System.Globalization
+
+Public Class Program
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseUnixEpoch.vb:3]: I could not find a test case like:` Dim asd As Date = New Date(1970, 1, 1)`

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM!
